### PR TITLE
Optimize interpod affinity priority function

### PR DESCRIPTION
**Pull request by ahg-g**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Optimizes inter-pod affinity priority function. The idea is to aggregate the score at the topology <key, value> level rather than the node. This allows us to avoid iterating over all the nodes in the cluster every time the score is updated.

**Does this PR introduce a user-facing change?**:
```release-note
Optimize inter-pod affinity preferredDuringSchedulingIgnoredDuringExecution type, up to 4x in some cases.
```

nxXHMMrzDEQ

### CODE REVIEW

**Comment by Huang-Wei** [comment]
Neat! I will take a look.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go [hunk 1]
```diff
@@ -69,13 +72,19 @@ func (p *podAffinityPriorityMap) processTerm(term *v1.PodAffinityTerm, podDefini
 	if err != nil {
 		return err
 	}
+	if fixedNode.Labels == nil {
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/priorities/interpod_affinity.go]: Prefer to `len(fixedNode.Labels) == 0` to tolerate the case it's inited with an empty map.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go [hunk 2]
```diff
@@ -69,13 +72,19 @@ func (p *podAffinityPriorityMap) processTerm(term *v1.PodAffinityTerm, podDefini
...
+
 	match := priorityutil.PodMatchesTermsNamespaceAndSelector(podToCheck, namespaces, selector)
-	if match {
-		for i, node := range p.nodes {
-			if priorityutil.NodesHaveSameTopologyKey(node, fixedNode, term.TopologyKey) {
-				atomic.AddInt64(&p.counts[i], weight)
-			}
+	tpValue, tpValueExist := fixedNode.Labels[term.TopologyKey]
+	if match && tpValueExist {
+		p.mu.Lock()
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/priorities/interpod_affinity.go]: p.Lock() is more conventional.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go [hunk 3]
```diff
@@ -69,13 +72,19 @@ func (p *podAffinityPriorityMap) processTerm(term *v1.PodAffinityTerm, podDefini
 	if err != nil {
 		return err
 	}
+	if fixedNode.Labels == nil {
+		return nil
+	}
+
 	match := priorityutil.PodMatchesTermsNamespaceAndSelector(podToCheck, namespaces, selector)
-	if match {
-		for i, node := range p.nodes {
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/priorities/interpod_affinity.go]: This is the key why it can boost the performance. Thanks!

**Review by Huang-Wei — COMMENTED** [review]
Thanks, @ahg-g ! Just some nits.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/algorithm/priorities/interpod_affinity.go]: done

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/algorithm/priorities/interpod_affinity.go]: done.

File: pkg/scheduler/algorithm/priorities/interpod_affinity.go (hunk 3 above)
Comment (ahg-g) [reply pkg/scheduler/algorithm/priorities/interpod_affinity.go]: Indeed.

**Pull request merged**